proposals/proposals/nnnn-trailing-comma-tuples-arguments-conditions.md

```diff
@@ -0,0 +1,200 @@
+
+
+**Tuples** use are very close to arguments list and, although may not be so frequently used, it seems natural that they adopt trailing comma too.
+
+**Multiple conditions** in `if`, `guard` and `while` are also comma-separated list and add, remove, reorder and comment in/out are not uncommon practice during development.
+
+```swift
+if 
+   condition1,
+   condition2,
+//   condition3
+{ ❌ Cannot convert value of type '() -> ()' to expected condition type 'Bool'
+                    
+} ❌ Expected '{' after 'if' condition
+```
+### Code Generation
+
+**Plugins** and **Macros** have made it possible to generate code using swift and trailing comma would allow generate list of arguments and conditions without worrying about a special condition for the last element.
+
+### Code Diff
+
+A tangential motivation is that trailing comma makes version-control diffs cleaner.
+
+Without trailing comma:
+```diff
+foo(
+-  a: Int
++  a: Int,
++  b: Int
+)
+```
+With trailing comma:
+```diff
+foo(
+  a: Int,
++  b: Int,
+)
+```
+
+> [!NOTE]  
+> A similar proposal was [rejected](https://forums.swift.org/t/rejected-se-0084-allow-trailing-commas-in-parameter-lists-and-tuples/2777) back in 2016 for Swift 3. It's been 8 years since that, the swift language has evolved a lot, some changes highlighted above as motivation, and the code style that "puts the terminating right parenthesis on a line following the arguments to that call" has been widely adopted by community, swift standard library codebase, swift-format, docc documentation and Xcode. Therefore, not encourage or endorse this code style doesn't hold true anymore nor is a reason for rejection.
+
+## Proposed solution
+
+This proposal adds support for trailing comma to:
+
+### Arguments and Parameters
+
+Including declaration and call of initializers, functions and enum case associated values.
+
+```swift
+
+func foo(
+    a: Int = 0, 
+    b: Int = 0, 
+) {
+}
+
+foo(
+    a: 1,
+    b: 2,
+)
+
+```
+
+### Tuples
+
+```swift
+(1, 2, 3,)
+```
+
+### Conditions
+
+Including `if`, `guard` and `while`.
+
+```swift
+if 
+   condition1,
+   condition2,
+   condition3,
+{
+                    
+}
+```
+
+## Source compatibility
+
+This change is purely additive and has no impact on existing code.
+
+## Future directions
+
+### Allow trailing comma anywhere there's a comma-separated list
+
+Although this proposal focuses on the most requested use cases for trailing comma, there's other places with comma-separated list and the restriction could be consistently lifted for all of these.
+
+#### Subclass and Protocol Conformance
+
+```swift
+class C2: C1, P1, P2, { } 
+```
+
+#### Generics
+
+```swift
+struct S<T1, T2,> where T1: P2, T2: P2, { }
+```
+#### Switch Case
+
+```swift
+switch number {
+    case 1, 2, 3,:
+        ...
+    default:
+        ...
+}
+```
```

Comment:
One of the justifications offered (which you should feel free to include in the text here) for expanding support for trailing commas is growing developer expectation, with many popular languages offering support. As folks have replied on the pitch thread, many languages either choose to stop at arrays/lists or support trailing commas basically everywhere that commas are used as delimiters.

Is there a principle or justification that underlies this proposal expanding support to include tuples and arguments, and control flow conditions _other than_ `switch`, but not these future directions? If so, can you articulate it here?

Reply:
I think I'm pro adding everywhere that is possible but I chose the ones that I personally considered more demanded for scope reason. I did considered add more during the implementation review but @ahoppen rightly [pointed](https://github.com/apple/swift-syntax/pull/2515#issuecomment-2006951192) that the review had already dragged too long so we went with just the proposed here. And to be honest, this has been more time-consuming than I expected so if possible I would prefer to not expand the proposal. If this is approved and there is a desire to expand, I would be willing to make another proposal expanding to more places.